File: app/src/main/java/com/skyd/imomoe/util/dlna/dmc/control/BaseServiceExecutor.java
package com.skyd.imomoe.util.dlna.dmc.control;


import android.os.Handler;
import android.os.Looper;

import androidx.annotation.Nullable;

import com.skyd.imomoe.util.dlna.dmc.Utils;
import com.skyd.imomoe.util.dlna.dmc.action.GetBrightness;
import com.skyd.imomoe.util.dlna.dmc.action.SetBrightness;

import org.fourthline.cling.controlpoint.ActionCallback;
import org.fourthline.cling.controlpoint.ControlPoint;
import org.fourthline.cling.model.action.ActionInvocation;
import org.fourthline.cling.model.message.UpnpResponse;
import org.fourthline.cling.model.meta.Service;
import org.fourthline.cling.support.avtransport.callback.GetMediaInfo;
import org.fourthline.cling.support.avtransport.callback.GetPositionInfo;
import org.fourthline.cling.support.avtransport.callback.GetTransportInfo;
import org.fourthline.cling.support.avtransport.callback.Pause;
import org.fourthline.cling.support.avtransport.callback.Play;
import org.fourthline.cling.support.avtransport.callback.Seek;
import org.fourthline.cling.support.avtransport.callback.SetAVTransportURI;
import org.fourthline.cling.support.avtransport.callback.Stop;
import org.fourthline.cling.support.model.MediaInfo;
import org.fourthline.cling.support.model.PositionInfo;
import org.fourthline.cling.support.model.TransportInfo;
import org.fourthline.cling.support.renderingcontrol.callback.GetMute;
import org.fourthline.cling.support.renderingcontrol.callback.GetVolume;
import org.fourthline.cling.support.renderingcontrol.callback.SetMute;
import org.fourthline.cling.support.renderingcontrol.callback.SetVolume;

/**
 *
 */
abstract class BaseServiceExecutor {
    private final ControlPoint mControlPoint;
    private final Service<?, ?> mService;
    private final Handler mHandler = new Handler(Looper.getMainLooper());

    protected BaseServiceExecutor(ControlPoint controlPoint, Service<?, ?> service) {
        mControlPoint = controlPoint;
        mService = service;
    }

    protected Service<?, ?> getService() {
        return mService;
    }

    protected boolean invalidServiceAction(String actionName) {
        return mService == null || mService.getAction(actionName) == null;
    }

    protected void execute(ActionCallback actionCallback) {
        mControlPoint.execute(actionCallback);
    }

    protected final <T> void notifySuccess(@Nullable IServiceAction.IServiceActionCallback<T> listener, T t) {
        if (listener != null) notify(() -> listener.onSuccess(t));
    }

    protected final void notifyFailure(@Nullable IServiceAction.IServiceActionCallback<?> listener, String errMsg) {
        if (listener != null) notify(() -> listener.onFailed(errMsg != null ? errMsg : "error"));
    }

    private void notify(Runnable runnable) {
        if (Looper.myLooper() != Looper.getMainLooper()) {
            mHandler.post(runnable);
        } else {
            runnable.run();
        }
    }

    // ---------------------------------------------------------------------------------------------------------
    // Implement
    // ---------------------------------------------------------------------------------------------------------
    static final class AVServiceExecutorImpl extends BaseServiceExecutor implements IServiceAction.IAVServiceAction {

        public AVServiceExecutorImpl(ControlPoint controlPoint, Service<?, ?> service) {
            super(controlPoint, service);
        }

        @Override
        public void cast(IServiceAction.IServiceActionCallback<String> listener, String uri, String metadata) {
            if (invalidServiceAction("SetAVTransportURI")) return;

            execute(new SetAVTransportURI(getService(), uri, metadata) {
                @Override
                public void success(final ActionInvocation invocation) {
                    notifySuccess(listener, uri);
                }

                @Override
                public void failure(final ActionInvocation invocation, final UpnpResponse operation, final String defaultMsg) {
                    notifyFailure(listener, defaultMsg);
                }
            });
        }

        @Override
        public void play(IServiceAction.IServiceActionCallback<Void> listener) {
            if (invalidServiceAction("Play")) return;

            execute(new Play(getService()) {
                @Override
                public void success(final ActionInvocation invocation) {
                    notifySuccess(listener, null);
                }

                @Override
                public void failure(final ActionInvocation invocation, final UpnpResponse operation, final String defaultMsg) {
                    notifyFailure(listener, defaultMsg);
                }
            });
        }

        @Override
        public void pause(IServiceAction.IServiceActionCallback<Void> listener) {
            if (invalidServiceAction("Pause")) return;

            execute(new Pause(getService()) {
                @Override
                public void success(final ActionInvocation invocation) {
                    notifySuccess(listener, null);
                }

                @Override
                public void failure(final ActionInvocation invocation, final UpnpResponse operation, final String defaultMsg) {
                    notifyFailure(listener, defaultMsg);
                }
            });
        }

        @Override
        public void stop(IServiceAction.IServiceActionCallback<Void> listener) {
            if (invalidServiceAction("Stop")) return;

            execute(new Stop(getService()) {
                @Override
                public void success(final ActionInvocation invocation) {
                    notifySuccess(listener, null);
                }

                @Override
                public void failure(final ActionInvocation invocation, final UpnpResponse operation, final String defaultMsg) {
                    notifyFailure(listener, defaultMsg);
                }
            });
        }

        @Override
        public void seek(IServiceAction.IServiceActionCallback<Long> listener, long position) {
            if (invalidServiceAction("Seek")) return;

            execute(new Seek(getService(), Utils.getStringTime(position)) {
                @Override
                public void success(final ActionInvocation invocation) {
                    notifySuccess(listener, position);
                }

                @Override
                public void failure(final ActionInvocation invocation, final UpnpResponse operation, final String defaultMsg) {
                    notifyFailure(listener, defaultMsg);
                }
            });
        }

        @Override
        public void getPositionInfo(IServiceAction.IServiceActionCallback<PositionInfo> listener) {
            if (invalidServiceAction("GetPositionInfo")) return;

            execute(new GetPositionInfo(getService()) {
                @Override
                public void received(ActionInvocation invocation, final PositionInfo positionInfo) {
                    notifySuccess(listener, positionInfo);
                }

                @Override
                public void failure(ActionInvocation invocation, UpnpResponse operation, String defaultMsg) {
                    notifyFailure(listener, defaultMsg);
                }
            });
        }

        @Override
        public void getMediaInfo(IServiceAction.IServiceActionCallback<MediaInfo> listener) {
            if (invalidServiceAction("GetMediaInfo")) return;

            execute(new GetMediaInfo(getService()) {
                @Override
                public void received(ActionInvocation invocation, final MediaInfo mediaInfo) {
                    notifySuccess(listener, mediaInfo);
                }

                @Override
                public void failure(ActionInvocation invocation, UpnpResponse operation, String defaultMsg) {
                    notifyFailure(listener, defaultMsg);
                }
            });
        }

        @Override
        public void getTransportInfo(IServiceAction.IServiceActionCallback<TransportInfo> listener) {
            if (invalidServiceAction("GetTransportInfo")) return;

            execute(new GetTransportInfo(getService()) {
                @Override
                public void received(ActionInvocation invocation, TransportInfo transportInfo) {
                    notifySuccess(listener, transportInfo);
                }

                @Override
                public void failure(ActionInvocation invocation, UpnpResponse operation, String defaultMsg) {
                    notifyFailure(listener, defaultMsg);
                }
            });
        }
    }

    // ---------------------------------------------------------------------------------------------------------
    // Implement
    // ---------------------------------------------------------------------------------------------------------
    static class RendererServiceExecutorImpl extends BaseServiceExecutor implements IServiceAction.IRendererServiceAction {

        public RendererServiceExecutorImpl(ControlPoint controlPoint, Service<?, ?> service) {
            super(controlPoint, service);
        }

        @Override
        public void setVolume(IServiceAction.IServiceActionCallback<Integer> listener, int volume) {
            if (invalidServiceAction("SetVolume")) return;

            execute(new SetVolume(getService(), volume) {
                @Override
                public void success(ActionInvocation invocation) {
                    notifySuccess(listener, volume);
                }

                @Override
                public void failure(ActionInvocation invocation, UpnpResponse operation, String defaultMsg) {
                    notifyFailure(listener, defaultMsg);
                }
            });
        }

        @Override
        public void getVolume(IServiceAction.IServiceActionCallback<Integer> listener) {
            if (invalidServiceAction("GetVolume")) return;

            execute(new GetVolume(getService()) {
                @Override
                public void received(ActionInvocation invocation, final int currentVolume) {
                    notifySuccess(listener, currentVolume);
                }

                @Override
                public void failure(ActionInvocation invocation, UpnpResponse operation, String defaultMsg) {
                    notifyFailure(listener, defaultMsg);
                }
            });
        }

        @Override
        public void setMute(IServiceAction.IServiceActionCallback<Boolean> listener, boolean mute) {
            if (invalidServiceAction("SetMute")) return;

            execute(new SetMute(getService(), mute) {
                @Override
                public void success(ActionInvocation invocation) {
                    notifySuccess(listener, mute);
                }

                @Override
                public void failure(ActionInvocation invocation, UpnpResponse operation, String defaultMsg) {
                    notifyFailure(listener, defaultMsg);
                }
            });
        }

        @Override
        public void isMute(IServiceAction.IServiceActionCallback<Boolean> listener) {
            if (invalidServiceAction("GetMute")) return;

            execute(new GetMute(getService()) {
                @Override
                public void received(ActionInvocation invocation, final boolean currentMute) {
                    notifySuccess(listener, currentMute);
                }

                @Override
                public void failure(ActionInvocation invocation, UpnpResponse operation, String defaultMsg) {
                    notifyFailure(listener, defaultMsg);
                }
            });
        }

        @Override
        public void setBrightness(IServiceAction.IServiceActionCallback<Integer> listener, int percent) {
            if (invalidServiceAction("SetBrightness")) return;

            execute(new SetBrightness(getService(), percent) {
                @Override
                public void success(final ActionInvocation invocation) {
                    notifySuccess(listener, percent);
                }

                @Override
                public void failure(ActionInvocation invocation, UpnpResponse operation, String defaultMsg) {
                    notifyFailure(listener, defaultMsg);
                }
            });
        }

        @Override
        public void getBrightness(IServiceAction.IServiceActionCallback<Integer> listener) {
            if (invalidServiceAction("GetBrightness")) return;

            execute(new GetBrightness(getService()) {
                @Override
                public void received(ActionInvocation<?> invocation, final int brightness) {
                    notifySuccess(listener, brightness);
                }

                @Override
                public void failure(ActionInvocation invocation, UpnpResponse operation, String defaultMsg) {
                    notifyFailure(listener, defaultMsg);
                }
            });
        }
    }
}

File: app/src/main/java/com/skyd/imomoe/util/dlna/dmc/control/BaseServiceExecutor.kt
package com.skyd.imomoe.util.dlna.dmc.control

import android.os.Handler
import org.fourthline.cling.controlpoint.ControlPoint
import android.os.Looper
import com.skyd.imomoe.util.dlna.dmc.Utils
import org.fourthline.cling.controlpoint.ActionCallback
import com.skyd.imomoe.util.dlna.dmc.control.IServiceAction.IServiceActionCallback
import com.skyd.imomoe.util.dlna.dmc.control.IServiceAction.IAVServiceAction
import org.fourthline.cling.model.action.ActionInvocation
import org.fourthline.cling.model.message.UpnpResponse
import com.skyd.imomoe.util.dlna.dmc.control.IServiceAction.IRendererServiceAction
import org.fourthline.cling.support.renderingcontrol.callback.SetVolume
import org.fourthline.cling.support.renderingcontrol.callback.GetVolume
import org.fourthline.cling.support.renderingcontrol.callback.SetMute
import org.fourthline.cling.support.renderingcontrol.callback.GetMute
import com.skyd.imomoe.util.dlna.dmc.action.SetBrightness
import com.skyd.imomoe.util.dlna.dmc.action.GetBrightness
import com.skyd.imomoe.util.showToast
import org.fourthline.cling.model.meta.Service
import org.fourthline.cling.support.avtransport.callback.*
import org.fourthline.cling.support.model.MediaInfo
import org.fourthline.cling.support.model.PositionInfo
import org.fourthline.cling.support.model.TransportInfo

internal abstract class BaseServiceExecutor protected constructor(
    private val mControlPoint: ControlPoint,
    protected val service: Service<*, *>?
) {
    private val mHandler = Handler(Looper.getMainLooper())

    protected fun invalidServiceAction(actionName: String): Boolean {
        return service?.getAction(actionName) == null
    }

    protected fun execute(actionCallback: ActionCallback) {
        mControlPoint.execute(actionCallback)
    }

    protected fun <T> notifySuccess(listener: IServiceActionCallback<T>?, t: T) {
        if (listener != null) notify { listener.onSuccess(t) }
    }

    protected fun notifyFailure(listener: IServiceActionCallback<*>?, errMsg: String?) {
        if (listener != null) notify { listener.onFailed(errMsg ?: "error") }
    }

    private fun notify(runnable: Runnable) {
        if (Looper.myLooper() != Looper.getMainLooper()) {
            mHandler.post(runnable)
        } else {
            runnable.run()
        }
    }

    // ---------------------------------------------------------------------------------------------------------
    // Implement
    // ---------------------------------------------------------------------------------------------------------
    internal class AVServiceExecutorImpl(controlPoint: ControlPoint, service: Service<*, *>?) :
        BaseServiceExecutor(controlPoint, service), IAVServiceAction {
        override fun cast(listener: IServiceActionCallback<String>?, uri: String, metadata: String) {
            if (invalidServiceAction("SetAVTransportURI")) return
            execute(object : SetAVTransportURI(service, uri, metadata) {
                override fun success(invocation: ActionInvocation<out Service<*, *>>?) {
                    notifySuccess(listener, uri)
                }

                override fun failure(
                    invocation: ActionInvocation<out Service<*, *>>?,
                    operation: UpnpResponse?,
                    defaultMsg: String?
                ) {
                    notifyFailure(listener, defaultMsg)
                }
            })
        }

        override fun play(listener: IServiceActionCallback<Void?>?) {
            if (invalidServiceAction("Play")) return
            execute(object : Play(service) {
                override fun success(invocation: ActionInvocation<out Service<*, *>>?) {
                    notifySuccess(listener, null)
                }

                override fun failure(
                    invocation: ActionInvocation<out Service<*, *>>?,
                    operation: UpnpResponse?,
                    defaultMsg: String?
                ) {
                    notifyFailure(listener, defaultMsg)
                }
            })
        }

        override fun pause(listener: IServiceActionCallback<Void?>?) {
            if (invalidServiceAction("Pause")) return
            execute(object : Pause(service) {
                override fun success(invocation: ActionInvocation<out Service<*, *>>?) {
                    notifySuccess(listener, null)
                }

                override fun failure(
                    invocation: ActionInvocation<out Service<*, *>>?,
                    operation: UpnpResponse?,
                    defaultMsg: String?
                ) {
                    notifyFailure(listener, defaultMsg)
                }
            })
        }

        override fun stop(listener: IServiceActionCallback<Void?>?) {
            if (invalidServiceAction("Stop")) return
            execute(object : Stop(service) {
                override fun success(invocation: ActionInvocation<out Service<*, *>>?) {
                    notifySuccess(listener, null)
                }

                override fun failure(
                    invocation: ActionInvocation<out Service<*, *>>?,
                    operation: UpnpResponse?,
                    defaultMsg: String?
                ) {
                    notifyFailure(listener, defaultMsg)
                }
            })
        }

        override fun seek(listener: IServiceActionCallback<Long>?, position: Long) {
            if (invalidServiceAction("Seek")) return
            execute(object : Seek(service, Utils.getStringTime(position)) {
                override fun success(invocation: ActionInvocation<out Service<*, *>>?) {
                    notifySuccess(listener, position)
                }

                override fun failure(
                    invocation: ActionInvocation<out Service<*, *>>?,
                    operation: UpnpResponse?,
                    defaultMsg: String?
                ) {
                    notifyFailure(listener, defaultMsg)
                }
            })
        }

        override fun getPositionInfo(listener: IServiceActionCallback<PositionInfo?>?) {
            if (invalidServiceAction("GetPositionInfo")) return
            execute(object : GetPositionInfo(service) {
                override fun received(
                    invocation: ActionInvocation<out Service<*, *>>?,
                    positionInfo: PositionInfo?
                ) {
                    notifySuccess(listener, positionInfo)
                }

                override fun failure(
                    invocation: ActionInvocation<out Service<*, *>>?,
                    operation: UpnpResponse?,
                    defaultMsg: String?
                ) {
                    notifyFailure(listener, defaultMsg)
                }
            })
        }

        override fun getMediaInfo(listener: IServiceActionCallback<MediaInfo?>?) {
            if (invalidServiceAction("GetMediaInfo")) return
            execute(object : GetMediaInfo(service) {
                override fun received(
                    invocation: ActionInvocation<out Service<*, *>>?,
                    mediaInfo: MediaInfo?
                ) {
                    notifySuccess(listener, mediaInfo)
                }

                override fun failure(
                    invocation: ActionInvocation<out Service<*, *>>?,
                    operation: UpnpResponse?,
                    defaultMsg: String?
                ) {
                    notifyFailure(listener, defaultMsg)
                }
            })
        }

        override fun getTransportInfo(listener: IServiceActionCallback<TransportInfo?>?) {
            if (invalidServiceAction("GetTransportInfo")) return
            execute(object : GetTransportInfo(service) {
                override fun received(
                    invocation: ActionInvocation<out Service<*, *>>?,
                    transportInfo: TransportInfo?
                ) {
                    notifySuccess(listener, transportInfo)
                }

                override fun failure(
                    invocation: ActionInvocation<out Service<*, *>>?,
                    operation: UpnpResponse?,
                    defaultMsg: String?
                ) {
                    notifyFailure(listener, defaultMsg)
                }
            })
        }
    }

    // ---------------------------------------------------------------------------------------------------------
    // Implement
    // ---------------------------------------------------------------------------------------------------------
    internal class RendererServiceExecutorImpl(
        controlPoint: ControlPoint,
        service: Service<*, *>?
    ) : BaseServiceExecutor(controlPoint, service), IRendererServiceAction {
        override fun setVolume(listener: IServiceActionCallback<Int>?, volume: Int) {
            if (invalidServiceAction("SetVolume")) return
            execute(object : SetVolume(service, volume.toLong()) {
                override fun success(invocation: ActionInvocation<out Service<*, *>>?) {
                    notifySuccess(listener, volume)
                }

                override fun failure(
                    invocation: ActionInvocation<out Service<*, *>>?,
                    operation: UpnpResponse?,
                    defaultMsg: String?
                ) {
                    notifyFailure(listener, defaultMsg)
                }
            })
        }

        override fun getVolume(listener: IServiceActionCallback<Int>?) {
            if (invalidServiceAction("GetVolume")) return
            execute(object : GetVolume(service) {
                override fun received(
                    actionInvocation: ActionInvocation<out Service<*, *>>?,
                    currentVolume: Int
                ) {
                    notifySuccess(listener, currentVolume)
                }

                override fun failure(
                    invocation: ActionInvocation<out Service<*, *>>?,
                    operation: UpnpResponse?,
                    defaultMsg: String?
                ) {
                    notifyFailure(listener, defaultMsg)
                }
            })
        }

        override fun setMute(listener: IServiceActionCallback<Boolean>?, mute: Boolean) {
            if (invalidServiceAction("SetMute")) return
            execute(object : SetMute(service, mute) {
                override fun success(invocation: ActionInvocation<out Service<*, *>>?) {
                    notifySuccess(listener, mute)
                }

                override fun failure(
                    invocation: ActionInvocation<out Service<*, *>>?,
                    operation: UpnpResponse?,
                    defaultMsg: String?
                ) {
                    notifyFailure(listener, defaultMsg)
                }
            })
        }

        override fun isMute(listener: IServiceActionCallback<Boolean>?) {
            if (invalidServiceAction("GetMute")) return
            execute(object : GetMute(service) {
                override fun received(
                    actionInvocation: ActionInvocation<out Service<*, *>>?,
                    currentMute: Boolean
                ) {
                    notifySuccess(listener, currentMute)
                }

                override fun failure(
                    invocation: ActionInvocation<out Service<*, *>>?,
                    operation: UpnpResponse?,
                    defaultMsg: String?
                ) {
                    notifyFailure(listener, defaultMsg)
                }
            })
        }

        override fun setBrightness(listener: IServiceActionCallback<Int>?, percent: Int) {
            if (invalidServiceAction("SetBrightness")) return
            try {
                execute(object : SetBrightness(service, percent.toLong()) {
                    override fun success(invocation: ActionInvocation<out Service<*, *>>?) {
                        notifySuccess(listener, percent)
                    }

                    override fun failure(
                        invocation: ActionInvocation<out Service<*, *>>?,
                        operation: UpnpResponse?,
                        defaultMsg: String?
                    ) {
                        notifyFailure(listener, defaultMsg)
                    }
                })
            } catch (e: IllegalArgumentException) {
                // service is null: ActionInvocation -> Action can not be null
                e.printStackTrace()
                e.message?.showToast()
            }
        }

        override fun getBrightness(listener: IServiceActionCallback<Int>?) {
            if (invalidServiceAction("GetBrightness")) return
            try {
                execute(object : GetBrightness(service) {
                    override fun received(
                        actionInvocation: ActionInvocation<out Service<*, *>>?,
                        brightness: Int
                    ) {
                        notifySuccess(listener, brightness)
                    }

                    override fun failure(
                        invocation: ActionInvocation<out Service<*, *>>?,
                        operation: UpnpResponse?,
                        defaultMsg: String?
                    ) {
                        notifyFailure(listener, defaultMsg)
                    }
                })
            } catch (e: IllegalArgumentException) {
                // service is null: ActionInvocation -> Action can not be null
                e.printStackTrace()
                e.message?.showToast()
            }
        }
    }
}